src/stacked_borrows.rs

```diff
@@ -283,32 +289,72 @@ impl Permission {
 impl<'tcx> Stack {
     /// Find the item granting the given kind of access to the given tag, and return where
     /// it is on the stack.
-    fn find_granting(&self, access: AccessKind, tag: SbTag) -> Option<usize> {
-        self.borrows
+    // TODO: Doc ok with Some(index) or None if unknown_bottom used
+    // Err if does not match
+    fn find_granting(
+        &self,
+        access: AccessKind,
+        tag: Option<SbTag>,
+        exposed_tags: &FxHashSet<SbTag>,
+    ) -> Result<Option<usize>, ()> {
+        let res = self
+            .borrows
             .iter()
             .enumerate() // we also need to know *where* in the stack
             .rev() // search top-to-bottom
             // Return permission of first item that grants access.
             // We require a permission with the right tag, ensuring U3 and F3.
-            .find_map(
-                |(idx, item)| {
-                    if tag == item.tag && item.perm.grants(access) { Some(idx) } else { None }
+            .find_map(|(idx, item)| {
+                match tag {
+                    Some(tag) if tag == item.tag && item.perm.grants(access) => Some(idx),
+                    None if exposed_tags.contains(&item.tag) => Some(idx),
```

Comment:
I think this will become cleaner if you handle the `None` tag first in a separate branch all the way at the top of the function. We don't need the `idx` in that case anyway, I think?

Also, the logic here is wrong -- we still need to check `item.perm.grants(access)`. We should probably have a testcase for that: if only SRO tags have been exposed, then a write access with a wildcard pointer should fail.